x/clp/keeper/rewards.go

```diff
@@ -0,0 +1,188 @@
+package keeper
+
+import (
+	"strings"
+
+	"github.com/Sifchain/sifnode/x/clp/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// BeginBlock prunes / auto-cancels the providers expired unlock requests.
+func (k Keeper) BeginBlock(ctx sdk.Context) {
+	params := k.GetParams(ctx)
+	for it := k.GetLiquidityProviderIterator(ctx); it.Valid(); it.Next() {
+		value := it.Value()
+		var lp types.LiquidityProvider
+		if len(value) <= 0 {
+			continue
+		}
+		err := k.cdc.Unmarshal(value, &lp)
+		if err != nil {
+			continue
+		}
+		k.PruneUnlockRecords(ctx, lp, params.LiquidityRemovalLockPeriod, params.LiquidityRemovalCancelPeriod)
+	}
+}
+
+// EndBlock distributes rewards to pools based on the current reward period configurations.
+func EndBlock(ctx sdk.Context, _ abci.RequestEndBlock, keeper Keeper) []abci.ValidatorUpdate {
+	params := keeper.GetParams(ctx)
+	pools := keeper.GetPools(ctx)
+	currentPeriod := keeper.GetCurrentRewardPeriod(ctx, params)
+	if currentPeriod != nil {
+		err := keeper.DistributeDepthRewards(ctx, currentPeriod, pools)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	return []abci.ValidatorUpdate{}
+}
+
+func (keeper Keeper) GetCurrentRewardPeriod(ctx sdk.Context, params types.Params) *types.RewardPeriod {
+	height := uint64(ctx.BlockHeight())
+	for _, period := range params.RewardPeriods {
+		if height >= period.StartBlock && height <= period.EndBlock {
+			return period
+		}
+	}
+	return nil
+}
+
+func (k Keeper) DistributeDepthRewards(ctx sdk.Context, period *types.RewardPeriod, pools []*types.Pool) error {
+	periodLength := period.EndBlock - period.StartBlock + 1
+	blockDistribution := period.Allocation.QuoUint64(periodLength)
```

Comment:
Can we move this calculation out of this function , and do it only once at period.StartBlock